### PR TITLE
Add WMCMSSWSubprocess and WMTiming metrics to WMArchvie  document

### LINKED ISSUE #11657: Propagate timestamp metrics from WMCore FWJR into WMArchive and MONIT
**Impact of the new feature**
Propagate timestamp metrics from CMSSW WM job into CMS monitoring

**Is your feature request related to a problem? Please describe.**
At the moment there is no clear timestamp metrics describing total wall clock time of WM jobs in CMS monitoring. Therefore, such request was placed in #11604. Even though it addresses how metrics should be set up in WM jobs we need an additional step to propagate them through WMArchive documents into MONIT.

**Describe the solution you'd like**
Propagate timestamp metrics from FWJR into WMArchive and MONIT:
- Adjust metrics related module to include these metrics from FWJR into WMArchive document
- Adjust WMArchvie schema to accept new metrics
- Adjust ES/OpenSearch schema/index in MONIT infrastructure
- Use new metrics in CMS monitoring dashboard(s)

**Describe alternatives you've considered**
None, we can continue using WM system as is and do not rely on this metrics

**Additional context**
Part of #11604

### PR DESCRIPTION
**Pull request by vkuznet**
Fixes #11657
Fixes the second part of #11660 and #11604

#### Status
ready

#### Description
Add WMCMSSWSubprocess and WMTiming metrics from FJR to WMArchive document

#### Is it backward compatible (if not, which system it affects?)
YES 

#### Related PRs
Complement to #11656 and #11665

### CODE REVIEW

**Comment by vkuznet** [comment]
@amaltaro , this PR contains CMSSW subprocess timing metrics. Should I add to this PR CMSSW performance one or would you like to have separate PR. In later case I'll probably need to rebase my second PR since this changes may be merged. While if we'll make appropriate changes here (with proper default) then we can speed-up the process of preparing WMArchive docs for both subprocess timing and performace metrics together. Since we provide defaults we may not need to have them in FJR document per-se. But we'll be ready with both on WMArchive side.

**Comment by amaltaro** [comment]
Valentin, what is the other PR that you are talking about?

Looking at the changes you provided in this PR, I would say it should be fixing https://github.com/dmwm/WMCore/issues/11657 instead of #11660, as you mentioned in the description.

We have a total of 3 GH issues involving time/cpu metrics, if possible, let's try to end up with 3 PRs closing them (I know we have one extra already with the test/json updates).

**Comment by vkuznet** [comment]
Alan, I was referring to CMSSW performance metrics which also need propagation to WMArchive, see https://github.com/dmwm/WMCore/pull/11663, or IO/Site metrics https://github.com/dmwm/WMCore/pull/11575/files and related WMArchive PRs to change the schema. Since in both cases it requires modification to WMCore/Serivce/WMArchive/DataMap to present them in final JSON that's why I asked should we add them in this PR or via separate PR.

**Comment by amaltaro** [comment]
For the CMSSW I/O metrics, I would rather have it in a separate pull request fixing the relevant GH issue (https://github.com/dmwm/WMCore/issues/11538). I guess there is already a pull request for the FJR part.

IMO, you could make a 2nd commit in the same PR providing the relevant changes for ArchiveDataReporter/WMArchive.

File: src/python/WMCore/Services/WMArchive/DataMap.py [hunk 1]
```diff
@@ -110,6 +111,12 @@
         # "GUID": '',
         # 'StageOutCommand': ''
     }],
+    'WMCMSSWSubprocess': {
```
Comment (amaltaro) [inline src/python/WMCore/Services/WMArchive/DataMap.py:114]: I feel like we should not be defining any default for these metrics, as it does not give a clear picture of whether it reads 0 or it is just the default value.
Is there any strong reason to set these defaults? Note that other attributes do not have any default!

**Review by amaltaro — CHANGES_REQUESTED** [review]
@vkuznet Valentin, please find a couple of comments along the code.

File: src/python/WMCore/Services/WMArchive/DataMap.py (hunk 1 above)
Comment (vkuznet) [reply src/python/WMCore/Services/WMArchive/DataMap.py]: ok, I removed defaults, usually I prefer to have defaults in place but it seems that they can be easily skipped if not provided and therefore not having them is ok too.

File: test/python/WMCore_t/Services_t/WMArchive_t/DataMap_t.py [hunk 2]
```diff
@@ -9,6 +9,11 @@
 SAMPLE_FWJR = {'fallbackFiles': [],
  'skippedFiles': [],
  'steps': {'cmsRun1': {'analysis': {},
+                       'WMCMSSWSubprocess': {'endTime': 1692718964.7409143,
```
[the remark this answers is not in the document]
Comment (vkuznet) [reply test/python/WMCore_t/Services_t/WMArchive_t/DataMap_t.py:12]: done

**Comment by vkuznet** [comment]
Alan, I added `WMTiming` metrics into this PR and it is ready to review. I also put in separate commit (https://github.com/dmwm/WMCore/pull/11692/commits/7d9a1527776d1165f72f9d6278a13a9faba0a8b7) changes to test json files to include `WMTiming` and `WMCMSSWSubprocess` metrics. This will allow to drop this PR https://github.com/dmwm/WMCore/pull/11659 in favor of this one.

**Review by amaltaro — COMMENTED** [review]
Having a second look at the DataMap module, I don't really understand what the difference is between `TOP_LEVEL_STEP_DEFAULT` and `STEP_DEFAULT`. I would say the former goes at the top level of the json, while the latter goes inside each step sub-dictionary (cmsRun1, cmsRun2, and the unneeded logArch, stageOut).

Said that, the expected behavior for these new metrics is:
* `WMTiming`: should be provided only once per WMArchive document (at the job top level json report)
* `WMCMSSWSubprocess`: should be provided within each step sub-dictionary of the json report (ideally only for CMSSW type, thus cmsRun1, cmsRun2, etc). Not at the top level json report, given that it's a cmsRun/CMSSW property.

To avoid bad surprises and further PR with follow up, I would suggest to take one of the FWJR  provided in this PR and pass it through the `createArchiveDoc` method:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/WMArchive/DataMap.py#L397C5-L397C22

to ensure that the final WMArchive document is as expected. 

Valentin, can you please do so and share the output file?

**Comment by amaltaro** [comment]
Nevermind! Let us get this in, run a test and then follow up with potential further fixes.

**Pull request merged**